spec/std/option_parser_spec.cr

```diff
@@ -438,4 +442,127 @@ describe "OptionParser" do
+    bar = false
+    sub2 = false
+    z = false
+    OptionParser.parse(args) do |opts|
+      opts.on("subcommand", "") do
+        subcommand = true
+        opts.on("--foo arg", "") { |v| foo = v }
+        opts.on("--bar", "") { bar = true }
+        opts.on("sub2", "") { sub2 = true }
+      end
+      opts.on("--verbose", "") { verbose = true }
+      opts.on("-z", "--baz", "") { z = true }
+    end
+
+    verbose.should be_true
+    subcommand.should be_true
+    foo.should be("1")
+    bar.should be_true
+    sub2.should be_true
+    z.should be_true
+  end
+
+  it "unregisters subcommands on call" do
+    foo = false
+    bar = false
+    baz = false
+    OptionParser.parse(%w(foo baz)) do |opts|
+      opts.on("foo", "") do
+        foo = true
+        opts.on("bar", "") { bar = true }
+      end
+      opts.on("baz", "") { baz = true }
+    end
+    foo.should be_true
+    bar.should be_false
+    baz.should be_false
+  end
+
+  it "handles subcommand --help well (top level)" do
+    help = nil
+    OptionParser.parse(%w(--help)) do |opts|
+      opts.banner = "Usage: foo"
+      opts.on("subcommand", "Subcommand Description") do
+        opts.on("-f", "--foo", "Foo") { }
+      end
+      opts.on("--verbose", "Verbose mode") { }
+      opts.on("--help", "Help") { help = opts.to_s }
+    end
+
+    help.should eq <<-USAGE
+      Usage: foo
+          subcommand                       Subcommand Description
+          --verbose                        Verbose mode
+          --help                           Help
+      USAGE
+  end
+
+  it "handles subcommand --help well (subcommand)" do
+    help = nil
+    OptionParser.parse(%w(subcommand --help)) do |opts|
+      opts.banner = "Usage: foo"
+      opts.on("subcommand", "Subcommand Description") do
+        opts.banner = "Usage: foo subcommand"
+        opts.on("-f", "--foo", "Foo") { }
+      end
+      opts.on("--verbose", "Verbose mode") { }
+      opts.on("--help", "Help") { help = opts.to_s }
+    end
+
+    help.should eq <<-USAGE
+      Usage: foo subcommand
+          --verbose                        Verbose mode
+          --help                           Help
+          -f, --foo                        Foo
+      USAGE
+  end
+
+  it "stops when asked" do
+    args = %w(--foo --stop --bar)
+    foo = false
+    bar = false
+    OptionParser.parse(args) do |opts|
+      opts.on("--foo", "") { foo = true }
+      opts.on("--bar", "") { bar = true }
+      opts.on("--stop", "") { opts.stop }
+      opts.unknown_args do |before, after|
+        before.should eq(%w())
+        after.should eq(%w(--bar))
+      end
+    end
+    foo.should be_true
+    bar.should be_false
+    args.should eq(%w(--bar))
+  end
+
+  it "can run a callback on every argument" do
+    args = %w(--foo file --bar)
+    foo = false
+    bar = false
+    OptionParser.parse(args) do |opts|
+      opts.on("--foo", "") { foo = true }
+      opts.on("--bar", "") { bar = true }
+      opts.every do |arg|
```

Comment:
I find it a bit confusing that calling `opts.stop` in `every` block behaves differently than in `on` block (/re whether the current arg is consumed). Maybe renaming it to `before_every` would help. Or `before_each`?

Reply:
yeah, before_each sounds good for me.